File: ios-swiftUI/Tday/Feature/Auth/LoginCredentialCoordinator.swift
import Foundation

@MainActor
final class LoginCredentialCoordinator {
    private var hasRequestedSystemCredential = false
    private var isRequestingSystemCredential = false

    func requestSavedCredentialIfAvailable(
        currentEmail: String = "",
        currentPassword: String,
        isCreatingAccount: Bool,
        isAuthLoading: Bool,
        credentialService: SystemCredentialServicing,
        login: (SystemCredential) async -> Bool
    ) async -> Bool {
        guard !isCreatingAccount,
              currentPassword.isEmpty,
              !isAuthLoading,
              !isRequestingSystemCredential,
              !hasRequestedSystemCredential else {
            return false
        }

        hasRequestedSystemCredential = true
        isRequestingSystemCredential = true
        defer { isRequestingSystemCredential = false }

        guard let credential = await credentialService.requestSavedCredential(preferredEmail: currentEmail) else {
            return false
        }

        return await login(credential)
    }
}

File: android-compose/app/src/main/java/com/ohmz/tday/compose/feature/auth/LoginCredentialCoordinator.kt
package com.ohmz.tday.compose.feature.auth

import android.content.Context
import com.ohmz.tday.compose.core.data.auth.SystemCredential

class LoginCredentialCoordinator {
    private var hasRequestedSystemCredential = false
    private var isRequestingSystemCredential = false

    suspend fun requestSavedCredentialIfAvailable(
        context: Context,
        currentEmail: String,
        currentPassword: String,
        isCreatingAccount: Boolean,
        isAuthLoading: Boolean,
        requestSavedCredential: suspend (Context, String?) -> SystemCredential?,
        login: suspend (SystemCredential) -> Boolean,
    ): Boolean {
        if (isCreatingAccount ||
            currentPassword.isNotEmpty() ||
            isAuthLoading ||
            isRequestingSystemCredential ||
            hasRequestedSystemCredential
        ) {
            return false
        }

        hasRequestedSystemCredential = true
        isRequestingSystemCredential = true
        try {
            val credential = requestSavedCredential(
                context,
                currentEmail.takeIf { it.isNotBlank() },
            ) ?: return false
            return login(credential)
        } finally {
            isRequestingSystemCredential = false
        }
    }
}
